### ISSUE TITLE
Speed up sol_verify

### ISSUE THREAD

**Issue by anirudhSK** [chipmunk-project/chipmunk#82]
On some benchmarks, verification takes a long time. 

For instance, on verifier_run.sk, attached here (suffix is txt instead of sk because of GitHub restrictions): [verifier_run.txt](https://github.com/chipmunk-project/chipmunk/files/3061670/verifier_run.txt), if you run:

sketch --bnd-inbits=10 verifier_run.sk, it takes 10+ hours to complete.

**Comment by anirudhSK** [comment]
We could use a full-blown verifier like Dafny, simplify the sketch before we feed it to sketch, or perform best-effort randomized testing at the risk of unsoundness.

**Comment by anirudhSK** [comment]
This and #60 might be related. One way to speed up sol_verify is to partially evaluate the sketch program before running the sketch compiler on it. Partial evaluation can also help us with visualization.

**Comment by taegyunkim** [comment]
I wonder whether simplification is a necessary step before feeding sketch to a verifier. Also, could sketch verification run time be reduced with this simplification? 

Anyway, yes, we could do probably two things to address your original point. 
1. Partial evaluation and/or simplification of sketch given hole value assignments.
2. Use verification tool to speed up verification.

**Comment by taegyunkim** [comment]
> On some benchmarks, verification takes a long time.
> 
> For instance, on verifier_run.sk, attached here (suffix is txt instead of sk because of GitHub restrictions): [verifier_run.txt](https://github.com/chipmunk-project/chipmunk/files/3061670/verifier_run.txt), if you run:
> 
> sketch --bnd-inbits=10 verifier_run.sk, it doesn't produce a result even after 10 hours.

Were there other benchmark you found similar issues? 

I wonder whether there is any commonality among the original program.

**Comment by anirudhSK** [comment]
Here's the simplest benchmark for verification:

harness void main(int x) {
  assert(x * 2 == x + x);
}

Even on this very simple benchmark, the verification time increases rapidly with increasing bnd-inbits. This means that even simplifying the verification sketch using partial evaluation is of limited value. A genuine verifier like Dafny is probably a better choice for sol_verify.

**Comment by taegyunkim** [comment]
https://rise4fun.com/Dafny/c88u

Dafny doesn't seem to prove equivalence of two arbitrary functions without any post-conditions. I don't think we can write formal post-conditions for all of our example specifications. 

Am I misunderstanding something? 

Please take a look at above link for examples. 

I also asked this on [stackoverflow](https://stackoverflow.com/questions/55661681/dafny-cant-prove-equivalence-of-methods), per guide in Dafny repo.

**Comment by anirudhSK** [comment]
What about just proving equivalence using statements of the first kind? For instance, assert pipeline_func(x) == program_func(x), where you don't create new variables, and hence probably don't need ensure statements.

**Comment by taegyunkim** [comment]
As far as I understand, we can't store any values in the functions. 

functions in Dafny can only have Expressions in function body. http://www.divms.uiowa.edu/~tinelli/classes/181/Papers/dafny-reference.pdf#page=74

Expressions can't have update statement (UpdateStmt) http://www.divms.uiowa.edu/~tinelli/classes/181/Papers/dafny-reference.pdf#page=102. 

Still, let me think through this more.

**Comment by anirudhSK** [comment]
This is really useful; thanks. What about writing the spec as a pure function (it will take some effort to transform it into one), and then using that pure function version of the spec as part of the ensures clause for the pipelined implementation?

**Comment by taegyunkim** [comment]
It might be possible if we can write a closed  form of the final value of variable that we want to check for all  conditions. 

https://rise4fun.com/Dafny/Pr1m
We probably convert Pipeline and Program in these ways, but somehow the verification fails. I wonder whether I made any mistake there. 

And one of the reasons that dafny is faster than Sketch is that it short-circuits verification using pre and post-conditions and don't look into other functions until it's really necessary to do so. 

From https://rise4fun.com/Dafny/tutorial/Guide and 
> The reason this happens is that Dafny "forgets" about the body of every method except the one it is currently working on.

**Comment by anirudhSK** [comment]
We could also look at Z3 directly. There's apparently an old sketch flag that generates a Z3 formula: https://lists.csail.mit.edu/pipermail/sketchusers/2018-November/000048.html It may not handle all the Sketch language constructs though.

**Comment by taegyunkim** [comment]
tl;dr: Decided to use z3 from sol_verify, and there are two complications. 

## Bnd-inbits
`sketch --beopt:writeSMT <pat-to-smt-file> <sketch-file>` this command also uses --bnd-inbits implying that setting the flag would increase the runtime as our current verification command does. 

However, for 
```C
harness void main(int x) {
  assert((x + x) == (x * 2));
}
```

It generates 
```SMT
(assert (forall ((x_1_1_0  Int  )) (let ((_n0 2 ))
(let ((_n1 x_1_1_0 ))
(let ((_n2 (*   _n1   _n0 ) ))
(let ((_n3 (+   _n1   _n1 ) ))
(let ((_n4 (=   _n3   _n2 ) ))
(implies  (and (>= x_1_1_0 0) (< x_1_1_0 32 ))    _n4  ))))))))
(check-sat)
(get-model)
(exit)
```

Where the constant 32 in the line starting with `(implies ...` just comes from --bnd-inbits. With --bnd-inbits=10 it will simply be 1024. 

So we can simply modify the file to look into our desired range of inputs. We eventually might want to look into sketch source code and find a flag which only transforms given sketch to a SMT file and not run its verifier. 

## Counter examples 

https://rise4fun.com/Z3/VbN8

This is the same SMT file except for the constant 2 (changed to 3) and other SMT pragmas to get the proof. 

Output of this program will contain following
```
(unit-resolution (unit-resolution @x95 (asserted a1) $x78) (mp ((_ quant-inst 1) $x122) @x130 $x121) false)))))))))))))))))))))))))))))))))))
```

Where quant-inst 1 means that the input 1 is a counter example, which is true (3 * 1 != 1 + 1). 

The original verifier_run.sk file's corresponding SMT file with unbounded input (it satisfies with 32 bit inputs) 
https://rise4fun.com/Z3/1DtK this gets `(_ quant-inst (- 1) (- 1) 7719 1236 0`

This way of getting counter examples require us to 
1. modify the generated smt file
2. use some weird regex match to parse the counter examples

**Comment by anirudhSK** [comment]
There's a Python API to Z3 that might make both 1 and 2 easier. Here's a description of the API: https://ericpony.github.io/z3py-tutorial/guide-examples.htm and here are some examples: https://github.com/Z3Prover/z3/tree/master/examples/python

**Comment by anirudhSK** [comment]
Closed by #92.
